File: 1456-FindTheCityWithTheSmallestNumberOfNeighborsAtAThresholdDistance/1456-FindTheCityWithTheSmallestNumberOfNeighborsAtAThresholdDistance.cpp
// Last updated: 12/8/2025, 11:26:12 PM
class Solution {
private:
    vector<int> minCityCount;
    vector<vector<int>> minDist;
    void floyd(int n, int distanceThreshold) {
        for (int k = 0; k < n; ++k) {
            for (int i = 0; i < n; ++i) {
                for (int j = 0; j < n; ++j) {
                    minDist[i][j] = min(minDist[i][j], minDist[i][k] + minDist[k][j]);
                    if (k == n - 1 && minDist[i][j] <= distanceThreshold && i != j) {
                        minCityCount[i]++;
                    }
                }
            }
        }
    }

public:
    int findTheCity(int n, vector<vector<int>>& edges, int distanceThreshold) {
        minCityCount.resize(n, 0);
        minDist.resize(n, vector<int>(n, 10005));
        for (const vector<int>& edge : edges) {
            minDist[edge[0]][edge[1]] = edge[2];
            minDist[edge[1]][edge[0]] = edge[2];
        }

        floyd(n, distanceThreshold);
        
        int city = -1;
        int num = n;
        for (int i = 0; i < n; ++i) {
            if (minCityCount[i] <= num) {
                city = i;
                num = minCityCount[i];
            }
        }
        return city;
    }
};

File: 1456-FindTheCityWithTheSmallestNumberOfNeighborsAtAThresholdDistance/1456-FindTheCityWithTheSmallestNumberOfNeighborsAtAThresholdDistance.py
# Last updated: 12/8/2025, 11:26:14 PM

class Solution:
    def findTheCity(
        self, n: int, edges: List[List[int]], distanceThreshold: int
    ) -> int:
        # Large value to represent infinity
        INF = int(1e9) + 7
        # Matrix to store shortest path distances from each city
        shortestPathMatrix = [[INF] * n for _ in range(n)]

        # Compute shortest paths from each city using Bellman-Ford algorithm
        for i in range(n):
            self.bellmanFord(n, edges, shortestPathMatrix[i], i)

        # Find the city with the fewest number of reachable cities within the distance threshold
        return self.getCityWithFewestReachable(
            n, shortestPathMatrix, distanceThreshold
        )

    # Bellman-Ford algorithm to find shortest paths from a source city
    def bellmanFord(
        self,
        n: int,
        edges: List[List[int]],
        shortestPathDistances: List[int],
        source: int,
    ) -> None:
        # Initialize distances from the source
        shortestPathDistances[:] = [float("inf")] * n
        shortestPathDistances[source] = 0  # Distance to source itself is zero

        # Relax edges up to n-1 times with early stopping
        for _ in range(n - 1):
            updated = False
            for start, end, weight in edges:
                if (
                    shortestPathDistances[start] != float("inf")
                    and shortestPathDistances[start] + weight
                    < shortestPathDistances[end]
                ):
                    shortestPathDistances[end] = (
                        shortestPathDistances[start] + weight
                    )
                    updated = True
                if (
                    shortestPathDistances[end] != float("inf")
                    and shortestPathDistances[end] + weight
                    < shortestPathDistances[start]
                ):
                    shortestPathDistances[start] = (
                        shortestPathDistances[end] + weight
                    )
                    updated = True
            if not updated:
                break  # Stop early if no updates

    # Determine the city with the fewest number of reachable cities within the distance threshold
    def getCityWithFewestReachable(
        self,
        n: int,
        shortestPathMatrix: List[List[int]],
        distanceThreshold: int,
    ) -> int:
        cityWithFewestReachable = -1
        fewestReachableCount = n

        # Count number of cities reachable within the distance threshold for each city
        for i in range(n):
            reachableCount = 0
            for j in range(n):
                if i == j:
                    continue  # Skip self
                if shortestPathMatrix[i][j] <= distanceThreshold:
                    reachableCount += 1

            # Update the city with the fewest reachable cities
            if reachableCount <= fewestReachableCount:
                fewestReachableCount = reachableCount
                cityWithFewestReachable = i
        return cityWithFewestReachable